File: src/bun.js/webcore/ArrayBufferSink.zig
const ArrayBufferSink = @This();

pub const JSSink = webcore.Sink.JSSink(@This(), "ArrayBufferSink");

bytes: bun.ByteList,
allocator: std.mem.Allocator,
done: bool = false,
signal: Signal = .{},
next: ?Sink = null,
streaming: bool = false,
as_uint8array: bool = false,

pub fn connect(this: *ArrayBufferSink, signal: Signal) void {
    bun.assert(this.reader == null);
    this.signal = signal;
}

pub fn start(this: *ArrayBufferSink, stream_start: streams.Start) bun.sys.Maybe(void) {
    this.bytes.clearRetainingCapacity();

    switch (stream_start) {
        .ArrayBufferSink => |config| {
            if (config.chunk_size > 0) {
                this.bytes.ensureTotalCapacityPrecise(this.allocator, config.chunk_size) catch
                    return .{ .err = Syscall.Error.oom };
            }

            this.as_uint8array = config.as_uint8array;
            this.streaming = config.stream;
        },
        else => {},
    }

    this.done = false;

    this.signal.start();
    return .success;
}

pub fn flush(_: *ArrayBufferSink) bun.sys.Maybe(void) {
    return .success;
}

pub fn flushFromJS(this: *ArrayBufferSink, globalThis: *JSGlobalObject, wait: bool) bun.sys.Maybe(JSValue) {
    if (this.streaming) {
        const value: JSValue = switch (this.as_uint8array) {
            true => jsc.ArrayBuffer.create(globalThis, this.bytes.slice(), .Uint8Array) catch .zero, // TODO: properly propagate exception upwards
            false => jsc.ArrayBuffer.create(globalThis, this.bytes.slice(), .ArrayBuffer) catch .zero, // TODO: properly propagate exception upwards
        };
        this.bytes.len = 0;
        if (wait) {}
        return .{ .result = value };
    }

    return .{ .result = JSValue.jsNumber(0) };
}

pub fn finalize(this: *ArrayBufferSink) void {
    this.destroy();
}

pub fn init(allocator: std.mem.Allocator, next: ?Sink) !*ArrayBufferSink {
    return bun.new(ArrayBufferSink, .{
        .bytes = bun.ByteList.empty,
        .allocator = allocator,
        .next = next,
    });
}

pub fn construct(
    this: *ArrayBufferSink,
    allocator: std.mem.Allocator,
) void {
    this.* = ArrayBufferSink{
        .bytes = bun.ByteList{},
        .allocator = allocator,
        .next = null,
    };
}

pub fn write(this: *@This(), data: streams.Result) streams.Result.Writable {
    if (this.next) |*next| {
        return next.writeBytes(data);
    }

    const len = this.bytes.write(this.allocator, data.slice()) catch {
        return .{ .err = Syscall.Error.oom };
    };
    this.signal.ready(null, null);
    return .{ .owned = len };
}
pub const writeBytes = write;
pub fn writeLatin1(this: *@This(), data: streams.Result) streams.Result.Writable {
    if (this.next) |*next| {
        return next.writeLatin1(data);
    }
    const len = this.bytes.writeLatin1(this.allocator, data.slice()) catch {
        return .{ .err = Syscall.Error.oom };
    };
    this.signal.ready(null, null);
    return .{ .owned = len };
}
pub fn writeUTF16(this: *@This(), data: streams.Result) streams.Result.Writable {
    if (this.next) |*next| {
        return next.writeUTF16(data);
    }
    const len = this.bytes.writeUTF16(this.allocator, @as([*]const u16, @ptrCast(@alignCast(data.slice().ptr)))[0..std.mem.bytesAsSlice(u16, data.slice()).len]) catch {
        return .{ .err = Syscall.Error.oom };
    };
    this.signal.ready(null, null);
    return .{ .owned = len };
}

pub fn end(this: *ArrayBufferSink, err: ?Syscall.Error) bun.sys.Maybe(void) {
    if (this.next) |*next| {
        return next.end(err);
    }
    this.signal.close(err);
    return .success;
}
pub fn destroy(this: *ArrayBufferSink) void {
    this.bytes.deinit(this.allocator);
    bun.destroy(this);
}
pub fn toJS(this: *ArrayBufferSink, globalThis: *JSGlobalObject, as_uint8array: bool) JSValue {
    if (this.streaming) {
        const value: JSValue = switch (as_uint8array) {
            true => jsc.ArrayBuffer.create(globalThis, this.bytes.slice(), .Uint8Array),
            false => jsc.ArrayBuffer.create(globalThis, this.bytes.slice(), .ArrayBuffer),
        };
        this.bytes.len = 0;
        return value;
    }

    defer this.bytes = bun.ByteList.empty;
    return ArrayBuffer.fromBytes(
        try this.bytes.toOwnedSlice(this.allocator),
        if (as_uint8array)
            .Uint8Array
        else
            .ArrayBuffer,
    ).toJS(globalThis, null);
}

pub fn endFromJS(this: *ArrayBufferSink, _: *JSGlobalObject) bun.sys.Maybe(ArrayBuffer) {
    if (this.done) {
        return .{ .result = ArrayBuffer.fromBytes(&[_]u8{}, .ArrayBuffer) };
    }

    bun.assert(this.next == null);
    this.done = true;
    this.signal.close(null);
    defer this.bytes = bun.ByteList.empty;
    return .{ .result = ArrayBuffer.fromBytes(
        bun.handleOom(this.bytes.toOwnedSlice(this.allocator)),
        if (this.as_uint8array)
            .Uint8Array
        else
            .ArrayBuffer,
    ) };
}

pub fn sink(this: *ArrayBufferSink) Sink {
    return Sink.init(this);
}

pub fn memoryCost(this: *const ArrayBufferSink) usize {
    // Since this is a JSSink, the NewJSSink function does @sizeOf(JSSink) which includes @sizeOf(ArrayBufferSink).
    return this.bytes.cap;
}

const std = @import("std");

const bun = @import("bun");
const Syscall = bun.sys;

const jsc = bun.jsc;
const ArrayBuffer = jsc.ArrayBuffer;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;

const webcore = bun.webcore;
const Sink = webcore.Sink;

const streams = webcore.streams;
const Signal = webcore.streams.Signal;

File: src/runtime/webcore/ArrayBufferSink.rs
use crate::webcore::streams::{self, SourceHandle};
use bun_collections::{ByteVecExt, VecExt};
use bun_jsc::HostReturn as _;
use bun_jsc::{ArrayBuffer, JSGlobalObject, JSType, JSValue};
use bun_sys as syscall;

// The "ArrayBufferSink" symbol-name concatenation lives in the `JsSinkAbi`
// impl in `Sink.rs` (see `array_buffer_sink_abi`).

#[derive(Default)]
pub struct ArrayBufferSink {
    pub(crate) bytes: Vec<u8>,
    // allocator field dropped — global mimalloc (non-AST crate, see PORTING.md §Allocators)
    pub(crate) done: bool,
    pub(crate) source: SourceHandle,
    pub(crate) streaming: bool,
    pub(crate) as_uint8array: bool,
}

impl ArrayBufferSink {
    pub(crate) fn start(&mut self, stream_start: &streams::Start) -> bun_sys::Result<()> {
        self.bytes.clear_retaining_capacity();

        if let streams::Start::ArrayBufferSink {
            chunk_size,
            as_uint8array,
            stream,
        } = *stream_start
        {
            if chunk_size > 0 {
                if self.bytes.try_reserve_exact(chunk_size as usize).is_err() {
                    return Err(syscall::Error::oom());
                }
            }

            self.as_uint8array = as_uint8array;
            self.streaming = stream;
        }

        self.done = false;

        self.source.start();
        Ok(())
    }

    pub fn flush(&mut self) -> bun_sys::Result<()> {
        Ok(())
    }

    pub(crate) fn flush_from_js(
        &mut self,
        global_this: &JSGlobalObject,
        _wait: bool,
    ) -> bun_sys::Result<JSValue> {
        if self.streaming {
            let value = if self.as_uint8array {
                ArrayBuffer::create::<{ JSType::Uint8Array }>(global_this, self.bytes.slice())
            } else {
                ArrayBuffer::create::<{ JSType::ArrayBuffer }>(global_this, self.bytes.slice())
            };
            self.bytes.clear();
            // Host return: empty ⇒ the exception `create` left pending.
            return Ok(value.or_pending_exception());
        }

        Ok(JSValue::js_number(0.0))
    }

    /// The sink a new `JSArrayBufferSink` wrapper owns; its `finalize`
    /// [`destroy`](Self::destroy)s it.
    pub(crate) fn construct() -> core::ptr::NonNull<Self> {
        bun_core::heap::into_raw_nn(Box::new(ArrayBufferSink {
            bytes: Vec::<u8>::default(),
            done: false,
            source: SourceHandle::default(),
            streaming: false,
            as_uint8array: false,
        }))
    }

    pub fn write(&mut self, data: &streams::Result) -> streams::result::Writable {
        let len = match self.bytes.write(data.slice()) {
            Ok(len) => len,
            Err(_) => return streams::result::Writable::Err(syscall::Error::oom()),
        };
        self.source.ready(None, None);
        streams::result::Writable::Owned(len as u64)
    }

    pub(crate) fn write_latin1(&mut self, data: &streams::Result) -> streams::result::Writable {
        let len = match self.bytes.write_latin1(data.slice()) {
            Ok(len) => len,
            Err(_) => return streams::result::Writable::Err(syscall::Error::oom()),
        };
        self.source.ready(None, None);
        streams::result::Writable::Owned(len as u64)
    }

    pub(crate) fn write_utf16(&mut self, data: &streams::Result) -> streams::result::Writable {
        let bytes = data.slice();
        // The caller guarantees the byte slice is u16-aligned and has even
        // length when the stream encoding is UTF-16. bytemuck checks both at
        // runtime.
        let utf16: &[u16] = bytemuck::cast_slice(bytes);
        let len = match self.bytes.write_utf16(utf16) {
            Ok(len) => len,
            Err(_) => return streams::result::Writable::Err(syscall::Error::oom()),
        };
        self.source.ready(None, None);
        streams::result::Writable::Owned(len as u64)
    }

    pub(crate) fn end(&mut self, err: Option<syscall::Error>) -> bun_sys::Result<()> {
        self.source.close(err);
        Ok(())
    }

    /// # Safety
    /// `this` is the allocation `js_construct` leaked into the JS wrapper, whose
    /// `__finalize` (the sole caller) frees it exactly once, here.
    pub(crate) unsafe fn destroy(this: *mut Self) {
        // SAFETY: reclaiming ownership drops `bytes` (Vec<u8> impls Drop) and
        // frees the box.
        drop(unsafe { bun_core::heap::take(this) });
    }

    pub(crate) fn end_from_js(
        &mut self,
        _global_this: &JSGlobalObject,
    ) -> bun_sys::Result<ArrayBuffer> {
        if self.done {
            return Ok(ArrayBuffer::from_bytes(&mut [], JSType::ArrayBuffer));
        }

        self.done = true;
        self.source.close(None);
        // `defer this.bytes = bun.Vec<u8>.empty` → take ownership, leave empty.
        let mut bytes = core::mem::take(&mut self.bytes);
        // Ownership transfers to JSC: the trait impl below converts the returned
        // `ArrayBuffer` with `to_js_unchecked`, which installs
        // `MarkedArrayBuffer_deallocator` (frees via `mi_free` on GC).
        let owned = bytes.to_owned_slice();
        Ok(ArrayBuffer::from_owned_bytes(
            owned,
            if self.as_uint8array {
                JSType::Uint8Array
            } else {
                JSType::ArrayBuffer
            },
        ))
    }

    pub(crate) fn memory_cost(&self) -> usize {
        // Since this is a JSSink, the NewJSSink function does @sizeOf(JSSink)
        // which includes @sizeOf(ArrayBufferSink).
        self.bytes.capacity() as usize
    }
}

// `JsSinkType` impl: routes the codegen `ArrayBufferSink__*` thunks (via
// `JSSink::<Self>::js_*`) into the inherent streaming methods above. Mirrors
// `Sink.JSSink(@This(), "ArrayBufferSink")`.
impl crate::webcore::sink::JsSinkType for ArrayBufferSink {
    const NAME: &'static str = "ArrayBufferSink";
    const HAS_CONSTRUCT: bool = true;
    const HAS_FLUSH_FROM_JS: bool = true;
    const START_TAG: Option<streams::StartTag> = Some(streams::StartTag::ArrayBufferSink);

    crate::impl_js_sink_forwarders!();

    fn finalize(this: bun_ptr::ThisPtr<Self>) {
        // SAFETY: trait contract — `this` is the wrapper's live sink, which
        // `construct` allocated for it alone; nothing uses it afterwards.
        unsafe { Self::destroy(this.as_ptr()) };
    }
    fn construct() -> core::ptr::NonNull<Self> {
        Self::construct()
    }
    fn end_from_js(&mut self, global: &JSGlobalObject) -> bun_sys::Result<JSValue> {
        match Self::end_from_js(self, global) {
            // Not `to_js`: its `mi_is_in_heap_region` probe would skip the
            // deallocator when the global allocator isn't mimalloc.
            bun_sys::Result::Ok(ab) => bun_sys::Result::Ok(match ab.to_js_unchecked(global) {
                Ok(v) => v,
                Err(_) => JSValue::ZERO,
            }),
            bun_sys::Result::Err(e) => bun_sys::Result::Err(e),
        }
    }
    fn source(&mut self) -> Option<&mut SourceHandle> {
        Some(&mut self.source)
    }
}
